Use qt5's mkDerivation in packages that otherwise crash

mmilata opened the pull request:
This attempts to fix several Qt applications that crash on startup with:
```
qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
```
A [crude script](https://gist.github.com/mmilata/9eeb216d01a560b9c98cebf5e9eb34fb) was used to find packages that:
* are called with `libsForQt5.callPackage` from `top-level/all-packages.nix`, and
* use `stdenv.mkDerivation`, but
* are not marked as `broken` nor do they refer to `wrapQtAppsHook`, and
* have binaries that crash with the message above.

This was followed with manual steps:
* Modify package file to use `mkDerivation` from argument list instead of `stdenv`, and
* build the package and check that some sort of window appears.

###### Motivation for this change

#65399

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Ensured that relevant documentation is up to date
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

----

Result of `nixpkgs-review pr 84673` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 package marked as broken and skipped:</summary>

  - bonfire
  - swiften
  - swiftformat
</details>
<details>
  <summary>34 package built:</summary>

  - aqemu
  - awesomebump
  - bibletime
  - bomi
  - calaos_installer
  - candle
  - caneda
  - colord-kde
  - dfasma
  - enyo-doom
  - firebird-emu
  - glogg
  - httraqt
  - iannix
  - ipe
  - kdeApplications.okteta
  - luckybackup
  - mindforger
  - mpc-qt
  - openbrf
  - phototonic
  - pro-office-calculator
  - qcomicbook
  - qmediathekview
  - qstopmotion
  - qt-box-editor
  - ricochet
  - rocket
  - swift-im
  - tensor
  - traverso
  - valentina
  - write_stylus
  - yabause
</details>

mmilata committed:
enyo-doom: use qt5's mkDerivation

mmilata committed:
httraqt: use qt5's mkDerivation

mmilata committed:
yabause: use qt5's mkDerivation

mmilata committed:
calaos_installer: use qt5's mkDerivation

mmilata committed:
caneda: use qt5's mkDerivation

mmilata committed:
write_stylus: use qt5's mkDerivation

mmilata committed:
valentina: use qt5's mkDerivation

mmilata committed:
traverso: use qt5's mkDerivation

mmilata committed:
swift-im: use qt5's mkDerivation

mmilata committed:
ricochet: use qt5's mkDerivation

mmilata committed:
qstopmotion: use qt5's mkDerivation

mmilata committed:
qmediathekview: use qt5's mkDerivation

mmilata committed:
qcomicbook: use qt5's mkDerivation

mmilata committed:
phototonic: use qt5's mkDerivation

mmilata committed:
openbrf: use qt5's mkDerivation

mmilata committed:
okteta: use qt5's mkDerivation

mmilata committed:
mindforger: use qt5's mkDerivation

mmilata committed:
dfasma: use qt5's mkDerivation

mmilata committed:
bomi: use qt5's mkDerivation

Wrap Qt program manually, remove makeWrapper from nativeBuildInputs.

mmilata committed:
bibletime: use qt5's mkDerivation

mmilata committed:
awesomebump: use qt5's mkDerivation

Wrap Qt program manually, remove makeWrapper from nativeBuildInputs.

mmilata committed:
aqemu: use qt5's mkDerivation

mmilata committed:
qt-box-editor: use qt5's mkDerivation

mmilata committed:
rocket: use qt5's mkDerivation

mmilata committed:
pro-office-calculator: use qt5's mkDerivation

mmilata committed:
iannix: use qt5's mkDerivation

mmilata committed:
glogg: use qt5's mkDerivation

mmilata committed:
firebird-emu: use qt5's mkDerivation

mmilata committed:
luckybackup: use qt5's mkDerivation

mmilata committed:
colord-kde: use qt5's mkDerivation

mmilata committed:
candle: use qt5's mkDerivation

mmilata committed:
tensor: use qt5's mkDerivation

mmilata commented on pkgs/applications/graphics/awesomebump/default.nix:
```diff
@@ -22,7 +22,7 @@ let
     '';
   };
-in stdenv.mkDerivation {
+in mkDerivation {
```
There's another `mkDerivation` above, not sure whether that also needs replacing.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/please-fix-all-packages-which-were-broken-by-the-qt-wrapping-changes/6444/28

das-g commented on pkgs/applications/graphics/awesomebump/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchgit, qtbase, qmake, makeWrapper, qtscript, flex, bison, qtdeclarative }:
+{ mkDerivation, lib, stdenv, fetchgit, qtbase, qmake, makeWrapper, qtscript, flex, bison, qtdeclarative }:
```
There's no other uses of `stdenv` in this function, so I guess it could be dropped from the arguments?

Or should it use `stdenv.lib` instead of the `lib` argument?

das-g commented on pkgs/tools/misc/colord-kde/default.nix:
```diff
@@ -1,11 +1,11 @@
-{ stdenv, lib, fetchurl
+{ mkDerivation, stdenv, lib, fetchurl
```
There's no other uses of `stdenv` in this function, so I guess it could be dropped from the arguments?

Or should it use `stdenv.lib` instead of the `lib` argument?

mmilata commented:
@das-g thank you for your attention! I've dropped `stdenv` from the arguments - looks better to me, not sure what's the right thing here, if there is such thing.

das-g reviewed (approved):
:+1: Change LGTM
:+1: Commits LGTM

worldofpeace reviewed (changes requested):
`bomi` does a conditional `wrapProgram` which will prevent it from being wrapped.
And that's because `wrapQtAppsHook` only wraps `ELF`, so it won't double wrap a script.

Other packages that have this problem are:
* `ipe`
* `awesomebump`

worldofpeace commented:
The part around here in the manual should be useful to you https://nixos.org/nixpkgs/manual/#qt-default-nix-co-3

mmilata commented:
@worldofpeace good point! In all three cases the hook ended up wrapping the .foo-wrapped binary so things worked out somewhat fine. Nevertheless having multiple levels of wrappers is ugly so I added `dontWrapQtApps = true;`, changed the wrapping function to the qt one and removed `makeWrapper` from `nativeBuildInputs` since it was no longer needed.

`bomi` remains broken though, as documented in https://github.com/NixOS/nixpkgs/issues/57532

worldofpeace reviewed (approved):
This LGTM. But there's no way I could even test this manually. Perhaps our other reviewer did this.

das-g commented:
> But there's no way I could even test this manually. Perhaps our other reviewer did this.

Me? I didn't test anything, I only looked at the diff and commit messages. I could run `nixpkgs-review pr` if that helps, but I guess whether stuff builds can also be tested by @GrahamcOfBorg, can't it?

The functionality of the build results I can't really test either, as I don't know the packaged applications. Maybe ping the respective package maintainers?

jtojnar commented:
I still think Qt's `mkDerivation` does more harm than good and should not have been introduced. Adding the dependencies (and `wrapQtAppsHook` for programs) should be sufficient. (There is also one minor improvement in debug flag handling but it does not outweigh the loss of compositionality. We should finish https://github.com/NixOS/nixpkgs/pull/71089 to get the debug flag handling to other qt setup hooks.)

worldofpeace commented:
@jtojnar I personally don't see much benefit in telling people to commit to something that isn't a committed practice, when it's mentioned in detail in our official documentation.
For what it's worth, I think any user of these commits would appreciate them committed right away.

@mmilata Can you rebase?

mmilata commented:
@worldofpeace rebased

worldofpeace commented:
I've backported this to 20.03, but I've yet to do it for 19.09 because there's conflicts.

mmilata commented:
Did the 19.09 in #85805 - gonna try building it now.

jtojnar commented:
Just to have it on the record, copying my rant from IRC:

> the fact that qt's `mkDerivation` should be removed was acknowledged by Thomas
> since it does not bring almost any benefits compared to `wrapQtAppsHook` I see no reason to continue using it, especially if it is to be deprecated
> the fact that documentation promotes it is unfortunate but it is a documentation bug that can be rectified

but since you already did the work and we do not want to make you redo it, I agreed with merging this for now:

> I guess we can merge it now and then do a removing sprint

worldofpeace commented:
Yep, thank you so much @mmilata for taking initiative to get these last fixes, I know people will appreciate it.

mmilata commented:
No need to merge this just to not upset me, I'm not too invested, just wanted to see if it's possible to fix some of the broken packages using a simple script:)

Anyway, #85805 built for me - had to drop `bibletime` and `write_stylus` since they'd require backporting more commits which is probably out of scope. Also `luckybackup` was added after 19.09 was forked and according to git log it probably never worked??
